### PR TITLE
AA WalletClientSigner Abstraction

### PR DESCRIPTION
**Pull request by joepegler**
# Summary

This PR facilitates the use of viems' `walletClient` as the signer while instantiating the smartWallet. 

Previously:

```
import {  WalletClientSigner } from "@alchemy/aa-core";
import { createSmartWalletClient } from "@biconomy/account"

const aaWalletClientSigner = new WalletClientSigner(viemWalletClient, "viem");
const smartWallet = await createSmartWalletClient({ chainId, signer:  aaWalletClientSigner})

```

Now:

```
import { createSmartWalletClient } from "@biconomy/account"
const smartWallet = await createSmartWalletClient({ signer:  viemWalletClient})
```

The chainId is now inferred from viem's walletClient

Related Issue: DEVX-405
Resolved: DEVX-422

## Change Type

- [ ] Bug Fix
- [ ] Refactor
- [x] New Feature
- [ ] Breaking Change
- [ ] Documentation Update
- [ ] Performance Improvement
- [x] Other

# Checklist

- [x] My code follows this project's style guidelines
- [x] I've reviewed my own code
- [x] I've added comments for any hard-to-understand areas
- [ ] I've updated the documentation if necessary
- [x] My changes generate no new warnings
- [x] I've added tests that prove my fix is effective or my feature works
- [x] All unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published

# Additional Information

This PR abstracts away AAs WalletClientSigner, so that viems' walletClient can be used as the signer instead.

We're doing work in the BiconomySmartAccountV2 before moving props into the constructor. This PR separates out types of the config passed into the static create call and types sent to the constructor.

### CODE REVIEW

File: packages/account/src/BiconomySmartAccountV2.ts [hunk 1]
```diff
@@ -141,18 +154,47 @@ export class BiconomySmartAccountV2 extends BaseSmartContractAccount {
...
-    const config = {
-      ...biconomySmartAccountConfig,
-    };
+    let chainId = biconomySmartAccountConfig.chainId;
+
+    // Signer needs to be initialised here before defaultValidationModule is set
+    if (biconomySmartAccountConfig.signer) {
+      const signer = biconomySmartAccountConfig.signer;
+      const isViemWalletClient = !(signer instanceof WalletClientSigner);
+      if (isViemWalletClient) {
```
Comment (livingrockrises) [inline packages/account/src/BiconomySmartAccountV2.ts:163]: What happens if it's not ViemWalletClient?

**Comment by livingrockrises** [comment]
Do add some high level summary besides ticket names. This is useful for external community

**Review by livingrockrises — APPROVED** [review]
review i

File: packages/account/src/BiconomySmartAccountV2.ts (hunk 1 above)
Comment (joepegler) [reply packages/account/src/BiconomySmartAccountV2.ts:163]: The regular flow, it passes through unchanged. Currently the default expectation is that the signer is AA's WalletClientSigner [(see here)](https://github.com/bcnmy/biconomy-client-sdk/blob/0f0a1b6b4b5d17b1d5bba13d285607881fcb8aed/packages/account/tests/account.spec.ts#L29)

File: packages/bundler/tests/index.d.ts [hunk 2]
```diff
@@ -2,8 +2,8 @@ import { Chain, Hex, PrivateKeyAccount, PublicClient, WalletClient } from "viem"
 import { WalletClientSigner } from "@alchemy/aa-core";
 
 interface WalletProps {
-  signer: WalletClientSigner;
-  walletClient: WalletClient;
+  alchemyWalletClientSigner: WalletClientSigner;
+  viemWallet: WalletClient;
```
[the remark this answers is not in the document]
Comment (joepegler) [reply packages/bundler/tests/index.d.ts:6]: No they're not, these are just explicitly named for the tests. Either can now be used as the signer. Prior to this PR only `alchemyWalletClientSigner` could be used, which was poor dx

**Comment by joepegler** [comment]
> Do add some high level summary besides ticket names. This is useful for external community

Done. Hope it's clearer now what the purpose of the task was

**Comment by livingrockrises** [comment]
Have we also thought about all the configurations we may need to pass on to createSmartWalletClient

because currently with "assumed ecdsa" default and active validation module, bundler just using bundlerUrl, paymaster using paymasterApiKey (only biconomy) may not be enough
If
- someone wants to "connect" or provide different module which needs more params or different kinds of signers 
- custom configuration when setting up bundlers like receipt timeouts
- custom configurations for paymaster
- any other flags we used in internal 'blocking methods' like buildUserOp (skiBundler estimation etc)

@AmanRaj1608 @VGabriel45 @joepegler

**Comment by joepegler** [comment]
> Have we also thought about all the configurations we may need to pass on to createSmartWalletClient
> 
> because currently with "assumed ecdsa" default and active validation module, bundler just using bundlerUrl, paymaster using paymasterApiKey (only biconomy) may not be enough If
> 
> * someone wants to "connect" or provide different module which needs more params or different kinds of signers
> * custom configuration when setting up bundlers like receipt timeouts
> * custom configurations for paymaster
> * any other flags we used in internal 'blocking methods' like buildUserOp (skiBundler estimation etc)
> 
> @AmanRaj1608 @VGabriel45 @joepegler

Yes. These configuration options can be passed into createSmartWalletClient in the same way as they were previously provided into BiconomySmartAccountV2Config.create. Nothing should have 'broken' with these new abstractions. The 'old' ways of customising options remain intact. 

in short:
```
await createSmartWalletClient({ signer: viemSmartWallet, bundler: new Bundler({ ...bundlerOptions }) });
```

etc will still work as expected

File: packages/account/src/utils/Types.ts [hunk 3]
```diff
@@ -90,32 +98,35 @@ type RequireAtLeastOne<T, Keys extends keyof T = keyof T> = Pick<T, Exclude<keyo
...
   },
   "defaultValidationModule" | "signer"
 >;
 
-type ConditionalBundlerProps = RequireAtLeastOne<
-  {
-    bundler: IBundler;
-    bundlerUrl: string;
-  },
-  "bundler" | "bundlerUrl"
```
[the remark this answers is not in the document]
Comment (joepegler) [reply packages/account/src/utils/Types.ts:103]: It's still there it's just moved further up the page

**Review by AmanRaj1608 — APPROVED** [review]

**Pull request merged**